### ISSUE TITLE
v2.3.0 (specifically #2310) breaks interactive (non-login) shell usage by prepending shims to PATH

### ISSUE THREAD

**Issue by Jmennius** [pyenv/pyenv#2389]
If I understand correctly, #2310 largely undoes what #1898 did.
Neither clearly specified a migration path (being breaking changes), causing issues for users...

The issue is that now, in interactive non-login shell `pyenv init -` modifies `PATH` (to add shims directory), which is undesirable and improper shell usage (what #1898 implemented correctly).
This is a breaking change. For example, it breaks `pipenv` shell if used with `pipenv shell --fancy` (which assumes _correct_ shell configuration) - by prepending shims directory which breaks virtualenv.

With #2310, is there a way to implement split configuration, i.e. have pyenv init not touch `PATH`? What is the correct configuration, what did I miss?

P.S. I do not mean to sound offensive, although I am a bit mad.

**Comment by native-api** [comment]
Prepending shims to PATH is essential for Pyenv's transparent switching mechanism. It was always in the suggested configuration. #1898 just moved prepending it to a different moment.

If you provide the necessary diagnostic information as per the issue template, showcasing your use case, we will likely be able to suggest something.

**Comment by native-api** [comment]
The thing is, the two-step configuration introduced in #1898 has proven to be too complicated and inconvenient for way too many users for comfort. So we reduced it back to 1-step but made it smarter to avoid duplicate PATH entries (which is why it was introduced in the first place).

We foresaw that this might break things for some advanced users with a complex custom setup (note that the suggested setup from 2.0-2.2 still works). Those however have shown to be a small minority -- so fixing it for regular users looked preferrable. As such, we decided to wait until someone with such an advanced use case so that we have some concrete usage info to design from.

**Comment by Jmennius** [comment]
Yes, I understand that.
My only issue is that there is no way to let pyenv do `init` without it modifying `PATH`.

I'll show my use case.
The setup is essentially what was suggested in 2.0-2.2:
I have this little script, which is sourced from `.zprofile` and `.zshrc`:
```zsh
# pyenv & pipenv setup
# this may be called twice - once for login shell (from .zprofile)
# and once for interactive shell (from .zshrc)

# assume there is no python environment if ~/.pyenv does not exist
[ -d $HOME/.pyenv ] || return

if [[ "$1" == "login" ]]; then
	# only execute in login shell - modifies PATH
	
	# prepend pyenv
	export PYENV_ROOT="$HOME/.pyenv"
	path=($PYENV_ROOT/bin $path)
	export PATH

	# not safe to executed with interactive shell - modifies the PATH
	eval "$(pyenv init --path --no-rehash)"

	export PIPENV_VENV_IN_PROJECT=1
	export PIPENV_SHELL_FANCY=1

elif [[ "$1" == "interactive" ]]; then
	# only useful in interactive shell

	eval "$(pyenv init - --no-rehash)"

	eval "$(_PIPENV_COMPLETE=zsh_source pipenv)"
fi
```
in `.zprofile`: `source ~/.config/zsh/pyenvrc login`.
and `.zshrc`: `source ~/.config/zsh/pyenvrc interactive`.

I use `pipenv`.
It has a `shell` (`pipenv shell`) command where you ask it to setup your shell for use with your python venv. `shell` has [compat](https://github.com/pypa/pipenv/blob/main/pipenv/shells.py#L99) and [normal](https://github.com/pypa/pipenv/blob/main/pipenv/shells.py#L86)/`fancy` modes.
Fancy here means that your shell is configured correctly - i.e. launching an interactive non-login shell should not modify `PATH`. The only thing pipenv does in this case is prepend venv bin directory to `PATH` and exec an interactive shell. I have fancy enabled globally via an envvar.
Compat is the default though. In compat pipenv will run a subshell and then source `activate`, which is a little bit ugly :)

So when I pulled pyenv today (had `v2.1.0-12-g5963dc4f` checked out before), `pipenv shell` has stopped working for me, but in a weird way which I couldn't figure out for a long time:
I had my python venv bin directory in `PATH`, but not python modules (site-packages of venv) - so I could invoke `ansible` installed in venv, but it would fail on importing something from `ansible` module.
The reason is that when I was invoking `python` - it used python from from pyenv shims, not from venv bin directory.
And venv 'activation' works by finding `pyvenv.cfg` in parent directory of the interpreter to understand that's a venv and add site-packages to `sys.path` - so venv python binary should be the one executed (directly?).

**Comment by native-api** [comment]
> 1. Can we ask users to exec a login shell? i.e. `exec $SHELL -l`

Since new terminal tabs/windows may be opening non-login shells -- no :slightly_smiling_face: . Too complicated for the poor users.

> What do you think about not moving shims to the beginning of the `PATH` if it's already there?

Considered that in https://github.com/pyenv/pyenv/issues/2085 as well as `--no-path`. [Having analyzed startup code for other language managers](https://github.com/pyenv/pyenv/issues/2085#issuecomment-1109178411), decided that moving would work in the widest variety of setups. So likewise postponed it until someone with a use case shows up.

**Comment by native-api** [comment]
> 2\. That's an interesting one. It's neither login nor interactive. Will look into it.

Worse. [In some distros (notably Debian-based and Fedora), Bash is configured to run `~/.bashrc` when run under SSH -- even non-interactively!](https://github.com/pyenv/pyenv/issues/2367)

**Comment by native-api** [comment]
> Considered that as well as `--no-path`.

I meant "considered when designing #2310". So if it works for you better, fine with implementing that, too.

**Comment by native-api** [comment]
> not moving shims to the beginning of the `PATH` if it's already there

I should have time to implement this next week. Looks like this will both fix your use case and address the corner cases.

**Comment by Jmennius** [comment]
Cool!
Sorry I didn't answer before, kinda didn't have the capacity...
I can offer this for `zsh`:
```diff
diff --git a/libexec/pyenv-init b/libexec/pyenv-init
index f3b0f38b..4693368a 100755
--- a/libexec/pyenv-init
+++ b/libexec/pyenv-init
@@ -147,6 +147,10 @@ function print_path() {
       echo 'set -eg PATH[$index]; end; set -e index'
       echo 'set -gx PATH '\'"${PYENV_ROOT}/shims"\'' $PATH'
       ;;
+    zsh )
+      echo 'if [[ -z ${path[(r)${PYENV_ROOT}/shims]} ]]; then path=(${PYENV_ROOT}/shims $path); fi'
+      echo 'export PATH'
+      ;;
     * )
       # Some distros (notably Debian-based) set Bash's SSH_SOURCE_BASHRC compilation option
       # that makes it source `bashrc` under SSH even when not interactive.

```
Seems to be working, I will test more.

**Comment by boldandbrad** [comment]
Not sure if this is entirely related, but in recent versions of pyenv I've noticed that when running `eval "$(pyenv init -)"` in my zshrc, anytime I run `which pyenv` I now get this atrocious output:

```
pyenv () {
        local command
        command="${1:-}" 
        if [ "$#" -gt 0 ]
        then
                shift
        fi
        case "$command" in
                (rehash | shell) eval "$(pyenv "sh-$command" "$@")" ;;
                (*) command pyenv "$command" "$@" ;;
        esac
}
```

instead of `/usr/local/bin/pyenv` as I would expect. I'm not sure how to fix this.

**Comment by native-api** [comment]
@boldandbrad this is by design. When the command you query is a shell function, Zsh's `which` prints the function's definition.

So see _both_ the shell function's definition and the program on PATH, run `which -a pyenv`.

**Comment by boldandbrad** [comment]
Thank you for explaining! I thought I was losing my mind. I hadn't noticed that behavior before and thought either my setup/config was wrong, or the tool had changed 😄

**Comment by jculpon** [comment]
I ran into a related issue moving a bunch of containerized tests from an old MBP used for stress testing (BigSur/latest) to a newer test environment which has a slightly different set of python installs and shims. Did the patch suggested by @native-api / @Jmennius get integrated in to a later 2.3.x or 2.6.x release?

Just trying to make sure a shell init issue I saw was config/2.3.x vs more modern pyenv related before trying something other than changing how the shell inits are handled on the old environment.
